glean-core/ios/Glean/Dispatchers.swift

```diff
@@ -0,0 +1,166 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/// This class manages background execution of Glean tasks.
+///
+/// This class makes use of the higher level `Operation` and `OperationQueue` API's in order to allow
+/// for observable background operation with the capabilities to pause, cancel, and resume tasks
+class Dispatchers {
+    /// This is the shared singleton access to the Glean Dispatchers
+    static let shared = Dispatchers()
+
+    // This struct is used for organizational purposes to keep the class constants in a single place
+    struct Constants {
+        static let logTag = "glean/Dispatchers"
+        static let maxQueueSize = 100
+    }
+
+    // This is the task queue that all Glean background operations will be executed on.  It is currently
+    // set to be a serial queue by setting the `maxConcurrentOperationsCount` to 1, but could allow for
+    // concurrent operations by increasing the parameter.
+    lazy var operationQueue: OperationQueue = {
+        var queue = OperationQueue()
+        queue.name = "Glean dispatch queue"
+        queue.maxConcurrentOperationCount = 1
+        return queue
+    }()
+
+    // When true, jobs will be queued and not ran until triggered by calling `flushQueuedInitialTasks()`
+    private var queueInitialTasks = AtomicBoolean(true)
+
+    // This array will hold the queued initial tasks that are launched before Glean is initialized
+    lazy var preInitOperations: [Operation] = [Operation]()
+
+    // When true, jobs will be run synchronously
+    var testingMode = false
+
+    // Don't let other instances be created, we only want singleton access through the static `shared`
+    // property
+    private init() {}
+
+    /// This function launches a background `Operation`
+    ///
+    /// - parameters:
+    ///   * block: The block of code to execute, as a Closure that accepts no arugments and returns Void
+    ///
+    /// This function is used throughout Glean in order to launch tasks in the background, typically
+    /// recording of metrics, uploading of pings, etc.  This function returns an `Operation` which can
+    /// be used in conjunction with a completion handler to perform work after the task is complete.
+    ///
+    /// For example:
+    /// ```
+    /// Dispatchers.shared.launch {
+    ///     // Do some work here
+    /// }.completionBlock = {
+    ///     print("Done!")
+    /// }
+    /// ```
+    ///
+    /// You could also use the returned `Operation` to await that specific task like this:
+    /// ```
+    /// let operation = Dispatchers.shared.launch {
```

Comment:
Wait, just assigning it to a variable automatically awaits the job to be finished?